### PR TITLE
ENH add Interpolant classes

### PR DESCRIPTION
**Pull request by beckermr**
This PR adds the interpolant classes from galsim.

### CODE REVIEW

File: tests/galsim_tests_config.yaml [hunk 1]
```diff
@@ -33,5 +33,5 @@ allowed_failures:
   - "module 'jax_galsim' has no attribute 'AutoCorrelate'"
   - "module 'jax_galsim' has no attribute 'AutoConvolve'"
   - "image.wrap"
-  - "unhashable type: 'ArrayImpl'"
-  - "'Image' object has no attribute 'bin'"
+  # - "unhashable type: 'ArrayImpl'"
+  # - "'Image' object has no attribute 'bin'"
```
Comment (beckermr) [inline tests/galsim_tests_config.yaml:37]:
```suggestion
  - "unhashable type: 'ArrayImpl'"
  - "'Image' object has no attribute 'bin'"
```

File: tests/conftest.py [hunk 2]
```diff
@@ -42,6 +42,11 @@ def pytest_pycollect_makemodule(module_path, path, parent):
     module = pytest.Module.from_parent(parent, path=module_path)
     # Overwrites the galsim module
     module.obj.galsim = __import__("jax_galsim")
+
+    # make sure tests of pickling work by having the jax_galsim module available
+    # in the galsim test helper functions
+    if hasattr(module.obj, "do_pickle"):
+        module.obj.do_pickle.__globals__["jax_galsim"] = __import__("jax_galsim")
```
Comment (beckermr) [inline tests/conftest.py:49]:
```suggestion
```

File: jax_galsim/gsparams.py [hunk 3]
```diff
@@ -10,7 +10,7 @@ class GSParams:
     minimum_fft_size: int = 128
     maximum_fft_size: int = 8192
     folding_threshold: float = 5.0e-3
-    stepk_minimum_hlr: float = 5
+    stepk_minimum_hlr: float = 5.0
```
Comment (beckermr) [inline jax_galsim/gsparams.py:13]:
```suggestion
    stepk_minimum_hlr: float = 5
```

File: jax_galsim/gsobject.py [hunk 4]
```diff
@@ -145,10 +145,10 @@ def __neg__(self):
...
 
     def __eq__(self, other):
-        is_same = self is other
-        is_same_class = type(other) is self.__class__
-        has_same_trees = self.tree_flatten() == other.tree_flatten()
-        return is_same or (is_same_class and has_same_trees)
+        return (self is other) or (
+            type(other) is self.__class__
+            and self.tree_flatten() == other.tree_flatten()
+        )
```
Comment (beckermr) [inline jax_galsim/gsobject.py:151]:
```suggestion
        is_same = self is other
        is_same_class = type(other) is self.__class__
        has_same_trees = self.tree_flatten() == other.tree_flatten()
        return is_same or (is_same_class and has_same_trees)
```

File: jax_galsim/interpolant.py [hunk 5]
```diff
@@ -0,0 +1,1564 @@
...
+    def urange(self):
+        """The maximum extent of the interpolant in Fourier space (in 2pi/pixels)."""
+        raise NotImplementedError(
+            "urange() is not implemented for interpolant type %s."
+            % self.__class__.__name__
+        )
+
+    # subclasses should implement __init__, _xval, _uval,
+    # _unit_integrals, _positive_flux, _negative_flux, urange, and xrange
+    # _xval and _uval should be static methods without the @staticmethod decorator
```
Comment (beckermr) [inline jax_galsim/interpolant.py:250]:
```suggestion
```

File: jax_galsim/interpolant.py [hunk 6]
```diff
@@ -0,0 +1,1564 @@
...
+class SincInterpolant(Interpolant):
+    # these magic numbers are from galsim itself via
+    # In [1]: import galsim
+    # In [2]: s = galsim.SincInterpolant()
+    # In [3]: s.positive_flux
+    # Out[3]: 3.18724409580418
+    # In [4]: s.negative_flux
+    # Out[4]: 2.18724409580418
+    _positive_flux = 3.18724409580418
+    _negative_flux = 2.18724409580418
```
Comment (beckermr) [inline jax_galsim/interpolant.py:347]: These likely depend on the k tolerance and so need to be fitting functions.

**Comment by beckermr** [comment]
OK friends! I have asked for a first-pass review on this PR. I am going to start implementing the actual interpolated image class before merging this one. I want to make sure the stuff in this PR will work best for the interpolated images classes themselves.

File: jax_galsim/interpolant.py [hunk 7]
```diff
@@ -0,0 +1,1653 @@
...
+            from galsim.deprecated import depr
+
+            depr("tol", 2.2, "gsparams=GSParams(kvalue_accuracy=tol)")
+            gsparams = GSParams(kvalue_accuracy=tol)
+        self._gsparams = GSParams.check(gsparams)
+
+    def _xval(self, x):
+        if jnp.ndim(x) > 1:
+            raise GalSimValueError("kval only takes scalar or 1D array values", x)
+
```
Comment (beckermr) [inline jax_galsim/interpolant.py:277]:
```suggestion
```

File: jax_galsim/interpolant.py [hunk 8]
```diff
@@ -0,0 +1,1678 @@
...
+    def withGSParams(self, gsparams=None, **kwargs):
+        """Create a version of the current interpolant with the given gsparams"""
+        if gsparams == self.gsparams:
+            return self
+        from copy import copy
+
+        ret = copy(self)
+        ret._gsparams = GSParams.check(gsparams, self.gsparams, **kwargs)
+        return ret
```
Comment (beckermr) [inline jax_galsim/interpolant.py:123]:
```suggestion
        if gsparams == self.gsparams:
            return self
        # Checking gsparams
        gsparams = GSParams.check(gsparams, self.gsparams, **kwargs)
        # Flattening the representation to instantiate a clean new object
        children, aux_data = self.tree_flatten()
        aux_data["gsparams"] = gsparams
        return self.tree_unflatten(aux_data, children)
```

**Comment by jecampagne** [comment]
Hello Matthew, 
I have tried to use your quintic interpolation code. In the following nb, I have C&P the code in the nb mentionned by @ismael-mendoza. I've also included some code that I've done for other things but it is limited. Although, I've difficulty to get right a test of obtaining a "high" resolution map from a "low" resolution one. I certainly, miss fit the input argument of the quintic interpolation function. 

[see a Collab nb here](https://colab.research.google.com/drive/1zNSAXwH2vT4xjhVT4toIT7uEqbiUscPY?usp=sharing) 

BTW if you impose float32 x variable init (cf. jnp.zero_like...) then the code crash if I use a JAX config with float64 variables.

**Comment by beckermr** [comment]
I cannot access the notebook. However the code in this pr matches galsim to pretty high precision so I'm guessing there isn't a bug in the code here.

**Comment by jecampagne** [comment]
I've open access, sorry.

**Comment by beckermr** [comment]
Fixed here: https://colab.research.google.com/drive/1tk8vPSiZac5PytN639dParpam9pR695w?usp=sharing

It was indeed the `jnp.zeros_like`! Thanks for catching the bug and the tip!

**Comment by jecampagne** [comment]
I'm glad to see that I have made a valuable x-check of the code, thanks having spent a bit to correct the nb.

File: jax_galsim/interpolant.py [hunk 9]
```diff
@@ -0,0 +1,1679 @@
...
+
+        def _high(x, n):
+            pix = jnp.pi * x
+            s = jnp.sin(pix)
+            sn = jnp.sin(pix / n)
+            return n * s * sn / (pix * pix)
+
+        msk_s = x <= 1e-4
+        msk_l = x <= n
+        val = jnp.piecewise(
```
Comment (ismael-mendoza) [inline jax_galsim/interpolant.py:1458]: I see you didn't specialize the `n` cases like in [Galsim](https://github.com/GalSim-developers/GalSim/blob/a1f0a11a24690a279abd7027b116ec3e4986643d/src/Interpolant.cpp#L579), do we loss significant speed this way?

**Review by ismael-mendoza — COMMENTED** [review]
This is a complex PR! As long as I can tell it's all correct, I carefully compared your implementation and constants with what is in `Interpolants.cpp` and `Interpolants.h` in Galsim. If there is anything I couldn't find or was unclear, I added a question below. 

Of course, there's a chance I missed something so I also pinged @jecampagne for a review. But I think I would be ok approving once the questions below are clear (since all tests are passing and I think you are additional waiting for all `InterpolatedImage` tests in #60 to pass as well before merging this one)

File: jax_galsim/bessel.py
Comment (ismael-mendoza) [inline jax_galsim/bessel.py:1]: might want to combine with `jax_galsim/core/bessel.py`?

File: jax_galsim/interpolant.py [hunk 10]
```diff
@@ -0,0 +1,1679 @@
...
+    def __init__(self, tol=None, gsparams=None):
+        if tol is not None:
+            from galsim.deprecated import depr
+
+            depr("tol", 2.2, "gsparams=GSParams(kvalue_accuracy=tol)")
+            gsparams = GSParams(kvalue_accuracy=tol)
+        self._gsparams = GSParams.check(gsparams)
+
+    @jax.jit
+    def _xval(x):
```
[the remark this answers is not in the document]
Comment (beckermr) [reply jax_galsim/interpolant.py:555]: Galsim uses Gary's original by default. You'd have to recompile it to get the other one.

File: jax_galsim/interpolant.py (hunk 9 above)
Comment (beckermr) [reply jax_galsim/interpolant.py:1458]: I didn't test that and plan to defer that work to future PRs. The goal here is to get something working.

File: jax_galsim/bessel.py
Comment (beckermr) [reply jax_galsim/bessel.py:1]: Nope. This is part of the official galsim upstream API.

File: jax_galsim/interpolant.py [hunk 11]
```diff
@@ -0,0 +1,1679 @@
...
+    @property
+    def ixrange(self):
+        """The total integral range of the interpolant.  Typically 2 * xrange."""
+        return 6
+
+
+@_wraps(_galsim.interpolant.Lanczos)
+@register_pytree_node_class
+class Lanczos(Interpolant):
+    # this data was generated in the dev notebook lanczos_interp_devel.ipynb
```
[the remark this answers is not in the document]
Comment (beckermr) [reply jax_galsim/interpolant.py:626]: ah I forgot to push it. Will add to this PR!

File: jax_galsim/interpolant.py [hunk 12]
```diff
@@ -0,0 +1,1679 @@
...
+            # these are constants so we do not propagate gradients
+            self._unit_integrals = jax.lax.stop_gradient(
+                (si(jnp.pi * (narr + 0.5)) - si(jnp.pi * (narr - 0.5))) / jnp.pi
+            )
+        return self._unit_integrals[:n]
+
+    def _comp_fluxes(self):
+        n = self.ixrange // 2 + 1
+        if n % 2 != 0:
+            n += 1
```
[the remark this answers is not in the document]
Comment (beckermr) [reply jax_galsim/interpolant.py:410]: The sinc function oscillates and so I wanted to integrate over an even number of periods to make things more stable. I will add a comment in the source.

File: jax_galsim/interpolant.py [hunk 13]
```diff
@@ -0,0 +1,1679 @@
...
+    # at include/Interpolant.h
+    _positive_flux = 1.1293413499280066555
+    _negative_flux = 0.1293413499280066555
+    _unit_integrals = jnp.array(
+        [
+            0.8724826228119177,
+            0.07332899380883082,
+            -0.010894097523532266,
+            0.0013237847222222375,
+        ],
```
[the remark this answers is not in the document]
Comment (beckermr) [reply jax_galsim/interpolant.py:542]: Ah woops. I got these from galsim itself but galsim computes them on the fly. I will add a comment and code snippet like the others.

**Comment by beckermr** [comment]
OK @ismael-mendoza - this one is ready to merge. I am still working on the interpolated image itself but I don't foresee a ton of changes here.

**Comment by beckermr** [comment]
I just rebased yes.

File: jax_galsim/interpolant.py [hunk 14]
```diff
@@ -0,0 +1,1681 @@
...
+@register_pytree_node_class
+class Interpolant:
+    def __init__(self):
+        raise NotImplementedError(
+            "The Interpolant base class should not be instantiated directly. "
+            "Use one of the subclasses instead, or use the `from_name` factory function."
+        )
+
+    @staticmethod
+    def from_name(name, tol=None, gsparams=None):
```
[the remark this answers is not in the document]
Comment (beckermr) [reply jax_galsim/interpolant.py:28]: No reason. The code base is not consistent on this, but I am leaving that to another PR.
